contracts/ROSCA.sol

```diff
@@ -0,0 +1,218 @@
+    * if there were no bid during the round, address selected at semi-random
+    *
+    * if currentRound = 0(ROSCA hasnt started), no winner is declared
+    */
+
+  function startRound() {
+    uint roundStartTime = startTime + (uint(currentRound)  * (uint(roundPeriodInDays) * 1 days));
+    if (now < roundStartTime || membersAddresses.length < minParticipants)
+      throw;
+
+    if (currentRound != 0) {
+      if (winnerAddress == 0) { // only true when there is no bidder in this round
+        // there is no bid in this round so find an unpaid address for this epoch
+        uint semi_random = now % membersAddresses.length;
+        for (uint i = 0; i < membersAddresses.length; i++) {
+          if(members[membersAddresses[(semi_random + i) % membersAddresses.length]].paid < EPOCH_TO_RUN)
+            winnerAddress = membersAddresses[semi_random + i];
+          break;
+        }
+      }
+      if(members[winnerAddress].paid == EPOCH_TO_RUN) throw;
+      members[winnerAddress].pendingWithdrawl += lowestBid;
+      members[winnerAddress].paid++;
+      LogRoundFundsReleased(winnerAddress, lowestBid);
+    }
+    if (currentRound < membersAddresses.length) {  // reset variables related to bidding
+      lowestBid = contributionSize * membersAddresses.length + 1;
+      winnerAddress = 0;
+
+      currentRound++;
+    }
+  }
+
+  /**
+   * Anyone not already a member of ROSCA can request to join and they'll be put into
+   * pendingJoinRequest until foreman accepts requests or ROSCA has started
+   */
+  function joinRequest()
+    onlyBeforeStart
+  {
+    // only put the request in the pending list if they are not in the ROSCA already
+    if (members[msg.sender].alive) throw;
+    pendingJoinRequest[msg.sender] = true;
+    LogParticipantApplied(msg.sender);
+  }
+
+  /**
+   * foreman can call this method to approve a join request by a participant
+   * once a requestor had been registered as member, the address will be taken out of pendingJoinRequest
+   **/
+  function acceptJoinRequest(address requestor)
+    onlyForeman
+    beforeStart
+  {
+    if (!pendingJoinRequest[requestor]) throw;
+    addMember(requestor);
+    LogParticipantApproved(requestor);
+    delete(pendingJoinRequest[requestor]); // take out the requestor's address in the pending list
+  }
+
+  /**
+   * Processes a periodic contribution from msg.sender ().
+   * Any excess funds will be withdrawable through withdraw().
+   */
+  function contribute()
+    payable
+  {
+    if (!members[msg.sender].alive) throw;
+    members[msg.sender].contributed += msg.value;
+    LogContributionMade(msg.sender, msg.value);
+  }
+
+  /**
+   * Registers a bid from msg.sender. If msg.sender has already won a round or bid is higher than lowestBid,
+   * this method will throw.
+   */
+  function bid(uint distrubtionAmountInWei) {
+    if (distrubtionAmountInWei >= lowestBid || members[msg.sender].paid >= EPOCH_TO_RUN || currentRound == 0
```

Comment:
* I think we want to prevent the user from asking a distribution that is < the MIN_DISTRIBUTION_RATIO, not >
* Consider breaking this into a few simpler lines. Specifically, it'd be nice to break the complex calculation into another line.
* Lot of codebases use 4-line indentation for long lines, to it's easy to see that a line was broken, e.g.
if (someVeryLongCondition && anotherCondition
    && yetAnotherCondition) {
  doSomething;
}

Would you be ok doing that?